From bb39a9ace3eece55e88caf4ef221550fb778edbd Mon Sep 17 00:00:00 2001
From: Richard Chang <richardycc@google.com>
Date: Fri, 13 Mar 2026 08:36:29 +0000
Subject: [PATCH 17/63] ANDROID: zram: Add ZRAM_ANDROID_IOC_PROCESS_PREFETCH
 ioctl

Introduce a new IOCTL command, ZRAM_ANDROID_IOC_PROCESS_PREFETCH, which
allows userspace to trigger the prefetching of pages from the backing
device into zram for a specific process.

When invoked, the zram driver scans the target process's page tables to
identify entries that are currently stored on the backing device. These
pages are then read from the backing device and written back into zram.
This mechanism is intended to reduce page fault swap-in latency by
restoring data from storage before it is accessed.

Bug: 495288884
Bug: 340137613
Change-Id: Ieebbf506663c09cb355636c8f5a81b7bd9f6d3fd
Signed-off-by: Richard Chang <richardycc@google.com>
(cherry picked from commit 160a4092db67397d85b790f102f375ae0e045257)
---
 drivers/block/zram/zram_drv.c   | 211 ++++++++++++++++++++++++++++++++
 drivers/block/zram/zram_drv.h   |   9 +-
 drivers/block/zram/zram_ioctl.c | 108 +++++++++++++---
 include/uapi/linux/zram_ioctl.h |   7 ++
 4 files changed, 313 insertions(+), 22 deletions(-)

diff --git a/drivers/block/zram/zram_drv.c b/drivers/block/zram/zram_drv.c
index af88a684d911..2155a8fafd4c 100644
--- a/drivers/block/zram/zram_drv.c
+++ b/drivers/block/zram/zram_drv.c
@@ -279,6 +279,16 @@ static inline void update_used_max(struct zram *zram,
 					  &cur_max, pages));
 }
 
+static bool zram_can_store_page(struct zram *zram)
+{
+	unsigned long alloced_pages;
+
+	alloced_pages = zs_get_total_pages(zram->mem_pool);
+	update_used_max(zram, alloced_pages);
+
+	return !zram->limit_pages || alloced_pages <= zram->limit_pages;
+}
+
 static inline void zram_fill_page(void *ptr, unsigned long len,
 					unsigned long value)
 {
@@ -450,6 +460,8 @@ static ssize_t idle_store(struct device *dev,
 static int read_from_zspool_raw(struct zram *zram, struct page *page,
 				u32 index);
 static int read_from_zspool(struct zram *zram, struct page *page, u32 index);
+static int read_from_bdev_sync(struct zram *zram, struct page *page, u32 index,
+			       unsigned long blk_idx);
 
 struct zram_wb_req {
 	unsigned long blk_idx;
@@ -903,6 +915,12 @@ static int zram_writeback_complete(struct zram *zram, struct zram_wb_req *req)
 	zram_set_handle(zram, index, req->blk_idx);
 	zram_set_flag(zram, index, ZRAM_WB);
 
+	/* Non-compressed writeback will decompress to PAGE_SIZE. */
+	if (!zram->compressed_wb) {
+		zram_set_obj_size(zram, index, PAGE_SIZE);
+		zram_set_flag(zram, index, ZRAM_HUGE);
+	}
+
 out:
 	zram_slot_unlock(zram, index);
 	return 0;
@@ -981,6 +999,169 @@ static struct zram_wb_req *zram_select_idle_req(struct zram_wb_ctl *wb_ctl)
 	return req;
 }
 
+static int zram_populate_table(struct zram *zram, struct page *page, u32 index)
+{
+	unsigned long handle;
+	void *src, *dst;
+	u32 size;
+
+	zram_slot_lock(zram, index);
+	/*
+	 * We release slot lock during zram_prefetch_slots, so slot can be
+	 * changed via slot_free(). To avoid the race, we need to check ZRAM_WB
+	 * again.
+	 */
+	if (!zram_test_flag(zram, index, ZRAM_WB)) {
+		zram_slot_unlock(zram, index);
+		return -EIO;
+	}
+
+	size = zram_get_obj_size(zram, index);
+
+	handle = zs_malloc(zram->mem_pool, size,
+			   GFP_NOWAIT | __GFP_HIGHMEM |
+			   __GFP_MOVABLE | __GFP_CMA);
+	if (IS_ERR_VALUE(handle)) {
+		zram_slot_unlock(zram, index);
+		return PTR_ERR((void *)handle);
+	}
+
+	if (!zram_can_store_page(zram)) {
+		zram_slot_unlock(zram, index);
+		zs_free(zram->mem_pool, handle);
+		return -ENOMEM;
+	}
+
+	src = kmap_local_page(page);
+	dst = zs_map_object(zram->mem_pool, handle, ZS_MM_WO);
+	memcpy(dst, src, size);
+	zs_unmap_object(zram->mem_pool, handle);
+	kunmap_local(src);
+
+	zram_clear_flag(zram, index, ZRAM_WB);
+	zram_release_bdev_block(zram, zram_get_handle(zram, index));
+	zram_set_handle(zram, index, handle);
+	atomic64_add(size, &zram->stats.compr_data_size);
+	if (zram_test_flag(zram, index, ZRAM_HUGE))
+		atomic64_inc(&zram->stats.huge_pages);
+	zram_slot_unlock(zram, index);
+
+	return 0;
+}
+
+static void zram_deferred_prefetch(struct work_struct *w)
+{
+	struct zram_rb_req *req = container_of(w, struct zram_rb_req, work);
+	struct page *page = bio_first_page_all(req->bio);
+	struct zram *zram = req->zram;
+	u32 index = req->index;
+
+	zram_populate_table(zram, page, index);
+
+	__free_page(page);
+	bio_put(req->bio);
+	kfree(req);
+}
+
+static void zram_prefetch_read_endio(struct bio *bio)
+{
+	struct zram_rb_req *req = bio->bi_private;
+	struct page *page = bio_first_page_all(bio);
+
+	if (bio->bi_status) {
+		__free_page(page);
+		bio_put(bio);
+		kfree(req);
+		return;
+	}
+
+	/*
+	 * Prefetch bdev page to zsmalloc_pool is sleepable.
+	 * We need to defer it to a preemptible context.
+	 */
+	INIT_WORK(&req->work, zram_deferred_prefetch);
+	queue_work(system_highpri_wq, &req->work);
+}
+
+static int zram_prefetch_from_bdev(struct zram *zram, struct page *page,
+				   u32 index, unsigned long blk_idx)
+{
+	struct zram_rb_req *req;
+	struct bio *bio;
+
+	req = kmalloc(sizeof(*req), GFP_KERNEL);
+	if (!req)
+		return -ENOMEM;
+
+	bio = bio_alloc(zram->bdev, 1, REQ_OP_READ, GFP_KERNEL);
+	if (!bio) {
+		kfree(req);
+		return -ENOMEM;
+	}
+
+	atomic64_inc(&zram->stats.bd_reads);
+
+	req->zram = zram;
+	req->index = index;
+	req->blk_idx = blk_idx;
+	req->bio = bio;
+
+	bio->bi_iter.bi_sector = blk_idx * (PAGE_SIZE >> 9);
+	bio->bi_private = req;
+	bio->bi_end_io = zram_prefetch_read_endio;
+
+	__bio_add_page(bio, page, PAGE_SIZE, 0);
+	submit_bio(bio);
+
+	return 0;
+}
+
+int zram_prefetch_slots(struct zram *zram, struct zram_pp_ctl *ctl)
+{
+	struct zram_pp_slot *pps;
+	int ret = 0;
+	u32 index = 0;
+	unsigned long blk_idx;
+	struct page *page = NULL;
+
+	while ((pps = select_pp_slot(ctl))) {
+		index = pps->index;
+		zram_slot_lock(zram, index);
+
+		if (!zram_test_flag(zram, index, ZRAM_PP_SLOT))
+			goto unlock_next;
+		if (!zram_test_flag(zram, index, ZRAM_WB))
+			goto unlock_next;
+
+		blk_idx = zram_get_handle(zram, index);
+		zram_slot_unlock(zram, index);
+
+		if (!page) {
+			page = alloc_page(GFP_KERNEL);
+			if (!page)
+				return -ENOMEM;
+		}
+
+		/* Read the page from backing device and restore to zram */
+		ret = zram_prefetch_from_bdev(zram, page, index, blk_idx);
+		if (ret)
+			break;
+
+		page = NULL;
+		release_pp_slot(zram, pps);
+		cond_resched();
+		continue;
+
+unlock_next:
+		zram_slot_unlock(zram, index);
+		release_pp_slot(zram, pps);
+	}
+
+	if (page)
+		__free_page(page);
+	return ret;
+}
+
 int zram_writeback_slots(struct zram *zram,
 			 struct zram_pp_ctl *ctl,
 			 struct zram_wb_ctl *wb_ctl)
@@ -1148,6 +1329,36 @@ static int parse_mode(char *val, u32 *mode)
 	return 0;
 }
 
+int scan_slot_for_prefetch(struct zram *zram, unsigned long index,
+			   struct zram_pp_ctl *ctl)
+{
+	struct zram_pp_slot *pps = NULL;
+
+	pps = kmalloc(sizeof(*pps), GFP_KERNEL);
+	if (!pps)
+		return -ENOMEM;
+
+	INIT_LIST_HEAD(&pps->entry);
+	pps->index = index;
+
+	zram_slot_lock(zram, index);
+	if (!zram_allocated(zram, index))
+		goto unlock_out;
+
+	if (!zram_test_flag(zram, index, ZRAM_WB))
+		goto unlock_out;
+
+	place_pp_slot(zram, ctl, pps);
+	pps = NULL;
+
+unlock_out:
+	zram_slot_unlock(zram, index);
+	kfree(pps);
+
+	return 0;
+}
+
+
 int scan_slots_for_writeback(struct zram *zram, u32 mode,
 			     unsigned long lo, unsigned long hi,
 			     struct zram_pp_ctl *ctl)
diff --git a/drivers/block/zram/zram_drv.h b/drivers/block/zram/zram_drv.h
index 9207abe810c0..67ef8be1df98 100644
--- a/drivers/block/zram/zram_drv.h
+++ b/drivers/block/zram/zram_drv.h
@@ -160,9 +160,6 @@ struct zram_pp_ctl {
 
 struct zram_pp_ctl *init_pp_ctl(void);
 void release_pp_ctl(struct zram *zram, struct zram_pp_ctl *ctl);
-int scan_slots_for_writeback(struct zram *zram, u32 mode,
-			     unsigned long lo, unsigned long hi,
-			     struct zram_pp_ctl *ctl);
 #endif
 
 #ifdef CONFIG_ZRAM_WRITEBACK
@@ -179,9 +176,15 @@ struct zram_wb_ctl {
 
 struct zram_wb_ctl *init_wb_ctl(struct zram *zram);
 void release_wb_ctl(struct zram_wb_ctl *wb_ctl);
+int scan_slots_for_writeback(struct zram *zram, u32 mode,
+			     unsigned long lo, unsigned long hi,
+			     struct zram_pp_ctl *ctl);
 int zram_writeback_slots(struct zram *zram,
 			 struct zram_pp_ctl *ctl,
 			 struct zram_wb_ctl *wb_ctl);
+int scan_slot_for_prefetch(struct zram *zram, unsigned long index,
+			   struct zram_pp_ctl *ctl);
+int zram_prefetch_slots(struct zram *zram, struct zram_pp_ctl *ctl);
 #endif
 
 #endif
diff --git a/drivers/block/zram/zram_ioctl.c b/drivers/block/zram/zram_ioctl.c
index 80ec70967a9f..c288cfa326f2 100644
--- a/drivers/block/zram/zram_ioctl.c
+++ b/drivers/block/zram/zram_ioctl.c
@@ -27,6 +27,7 @@
 struct zram_process_walk_private {
 	struct zram *zram;
 	struct zram_pp_ctl *pp_ctl;
+	unsigned int cmd;
 	u64 nr_remaining_pages;
 	unsigned long next_addr;
 };
@@ -58,6 +59,7 @@ static int zram_process_walker(pmd_t *pmd, unsigned long start,
 	struct zram_process_walk_private *private = walk->private;
 	struct zram *zram = private->zram;
 	struct zram_pp_ctl *pp_ctl = private->pp_ctl;
+	unsigned int cmd = private->cmd;
 	struct vm_area_struct *vma = walk->vma;
 	struct swap_info_struct *sis;
 	pte_t *ptep, pte;
@@ -99,10 +101,15 @@ static int zram_process_walker(pmd_t *pmd, unsigned long start,
 		if (unlikely(index >= nr_pages))
 			goto unlock_swap_device;
 
-		/* Use PAGE_WRITEBACK for single index */
-		scan_slots_for_writeback(zram, 0, index, index+1, pp_ctl);
-		if (private->nr_remaining_pages != NR_PAGES_UNLIMITED)
-			private->nr_remaining_pages--;
+		if (cmd == ZRAM_ANDROID_IOC_PROCESS_RANGE_WRITEBACK) {
+			/* Use PAGE_WRITEBACK for single index */
+			scan_slots_for_writeback(zram, 0, index, index+1,
+						 pp_ctl);
+			if (private->nr_remaining_pages != NR_PAGES_UNLIMITED)
+				private->nr_remaining_pages--;
+		} else if (cmd == ZRAM_ANDROID_IOC_PROCESS_PREFETCH) {
+			scan_slot_for_prefetch(zram, index, pp_ctl);
+		}
 
 unlock_swap_device:
 		put_swap_device(sis);
@@ -117,28 +124,36 @@ static const struct mm_walk_ops zram_walk_ops = {
 	.walk_lock = PGWALK_RDLOCK,
 };
 
-static int zram_ioctl_process_writeback_scan(struct zram *zram,
-	struct zram_android_ioc_process_range_writeback *ioc_prwb,
+static int zram_ioctl_process_scan(struct zram *zram, unsigned int cmd,
+	__aligned_u64 pidfd,
+	struct zram_android_ioc_process_range_writeback *prwb,
 	struct zram_pp_ctl *ctl)
 {
 	struct mm_struct *mm;
 	struct vm_area_struct *vma;
 	struct task_struct *task;
-	unsigned long start_addr = (unsigned long)ioc_prwb->start_addr;
+	unsigned long start_addr = 0;
+	u64 nr_remaining_pages = 0;
 	unsigned int f_flags;
 	int ret = 0;
 
 	struct zram_process_walk_private private = {
 		.zram = zram,
 		.pp_ctl = ctl,
-		.nr_remaining_pages = DIV_ROUND_UP_POW2(ioc_prwb->size,
-							PAGE_SIZE),
+		.cmd = cmd,
 	};
 
-	if (!private.nr_remaining_pages)
+	if (cmd == ZRAM_ANDROID_IOC_PROCESS_RANGE_WRITEBACK) {
+		start_addr = (unsigned long)prwb->start_addr;
+		nr_remaining_pages = DIV_ROUND_UP_POW2(prwb->size, PAGE_SIZE);
+	}
+
+	if (!nr_remaining_pages)
 		private.nr_remaining_pages = NR_PAGES_UNLIMITED;
+	else
+		private.nr_remaining_pages = nr_remaining_pages;
 
-	task = pidfd_get_task(ioc_prwb->pidfd, &f_flags);
+	task = pidfd_get_task(pidfd, &f_flags);
 	if (IS_ERR(task))
 		return PTR_ERR(task);
 
@@ -174,11 +189,13 @@ static int zram_ioctl_process_writeback_scan(struct zram *zram,
 			break;
 	}
 
-	if (ret > 0) {
-		ioc_prwb->next_addr = private.next_addr;
-		ret = 0;
-	} else {
-		ioc_prwb->next_addr = 0;
+	if (cmd == ZRAM_ANDROID_IOC_PROCESS_RANGE_WRITEBACK) {
+		if (ret > 0) {
+			prwb->next_addr = private.next_addr;
+			ret = 0;
+		} else {
+			prwb->next_addr = 0;
+		}
 	}
 
 	mmap_read_unlock(mm);
@@ -191,7 +208,7 @@ static int zram_ioctl_process_writeback_scan(struct zram *zram,
 }
 
 static int zram_ioctl_process_writeback(struct zram *zram,
-	struct zram_android_ioc_process_range_writeback *ioc_prwb)
+	struct zram_android_ioc_process_range_writeback *prwb)
 {
 	struct zram_pp_ctl *pp_ctl = NULL;
 	struct zram_wb_ctl *wb_ctl = NULL;
@@ -224,11 +241,13 @@ static int zram_ioctl_process_writeback(struct zram *zram,
 		goto clear_pp_ctl;
 	}
 
-	ret = zram_ioctl_process_writeback_scan(zram, ioc_prwb, pp_ctl);
+	ret = zram_ioctl_process_scan(zram,
+				      ZRAM_ANDROID_IOC_PROCESS_RANGE_WRITEBACK,
+				      prwb->pidfd, prwb, pp_ctl);
 	if (!ret)
 		ret = zram_writeback_slots(zram, pp_ctl, wb_ctl);
 
-	ioc_prwb->written_bytes = wb_ctl->processed_bytes;
+	prwb->written_bytes = wb_ctl->processed_bytes;
 
 	release_wb_ctl(wb_ctl);
 clear_pp_ctl:
@@ -239,6 +258,50 @@ static int zram_ioctl_process_writeback(struct zram *zram,
 	return ret;
 }
 
+static int zram_ioctl_process_prefetch(struct zram *zram,
+	struct zram_android_ioc_process_prefetch *ioc_prefetch)
+{
+	struct zram_pp_ctl *pp_ctl = NULL;
+	int ret;
+
+	/* Require CAP_SYS_NICE for influencing process performance. */
+	if (!capable(CAP_SYS_NICE))
+		return -EPERM;
+
+	guard(rwsem_read)(&zram->init_lock);
+	if (!init_done(zram))
+		return -EINVAL;
+
+	if (!zram->backing_dev)
+		return -ENODEV;
+
+	/*
+	 * Prefetch should preempt writeback for the same process to avoid
+	 * blocking the launch. However, Prefetch and Writeback work in
+	 * parallel could have a race of block index, even operate in different
+	 * processes. Currently, we do not permit concurrent post-processing
+	 * actions via pp_in_progress flag.
+	 */
+	if (atomic_xchg(&zram->pp_in_progress, 1))
+		return -EAGAIN;
+
+	pp_ctl = init_pp_ctl();
+	if (!pp_ctl) {
+		ret = -ENOMEM;
+		goto clear_pp_in_progress;
+	}
+
+	ret = zram_ioctl_process_scan(zram, ZRAM_ANDROID_IOC_PROCESS_PREFETCH,
+				      ioc_prefetch->pidfd, NULL, pp_ctl);
+	if (!ret)
+		ret = zram_prefetch_slots(zram, pp_ctl);
+
+	release_pp_ctl(zram, pp_ctl);
+clear_pp_in_progress:
+	atomic_set(&zram->pp_in_progress, 0);
+	return ret;
+}
+
 int zram_ioctl(struct block_device *bdev, blk_mode_t mode,
 	       unsigned int cmd, unsigned long arg)
 {
@@ -256,6 +319,13 @@ int zram_ioctl(struct block_device *bdev, blk_mode_t mode,
 
 		if (copy_to_user(argp, &prwb, sizeof(prwb)))
 			ret = -EFAULT;
+	} else if (cmd == ZRAM_ANDROID_IOC_PROCESS_PREFETCH) {
+		struct zram_android_ioc_process_prefetch ioc_prefetch;
+
+		if (copy_from_user(&ioc_prefetch, argp, sizeof(ioc_prefetch)))
+			return -EFAULT;
+
+		ret = zram_ioctl_process_prefetch(zram, &ioc_prefetch);
 	} else if (cmd == ZRAM_ANDROID_IOC_PROCESS_WRITEBACK) {
 		/* Legacy: map to unlimited process_range_writeback */
 		struct zram_android_ioc_data ioc_data;
diff --git a/include/uapi/linux/zram_ioctl.h b/include/uapi/linux/zram_ioctl.h
index 99519690f904..5838dbc809eb 100644
--- a/include/uapi/linux/zram_ioctl.h
+++ b/include/uapi/linux/zram_ioctl.h
@@ -48,6 +48,10 @@ struct zram_android_ioc_process_range_writeback {
 	__u64		written_bytes;
 };
 
+struct zram_android_ioc_process_prefetch {
+	__aligned_u64	pidfd;
+};
+
 #define ZRAM_ANDROID_IOC_MAGIC 0xBB
 
 /* Legacy API: keep it for compatibility */
@@ -57,5 +61,8 @@ struct zram_android_ioc_process_range_writeback {
 #define ZRAM_ANDROID_IOC_PROCESS_RANGE_WRITEBACK \
 	_IOWR(ZRAM_ANDROID_IOC_MAGIC, 2, struct zram_android_ioc_process_range_writeback)
 
+#define ZRAM_ANDROID_IOC_PROCESS_PREFETCH \
+	_IOW(ZRAM_ANDROID_IOC_MAGIC, 3, struct zram_android_ioc_process_prefetch)
+
 #endif /* _UAPI_LINUX_ZRAM_IOCTL_H */
 
-- 
2.50.1

